### PR TITLE
Feature/automatically generate definitions

### PR DESCRIPTION
**Pull request by TomHogervorst**
This PR addresses issues #23 , issue #24 and issue #15, and fixes the bug describes in issue #14 by creating a new trace set parameter map by default.

### CODE REVIEW

File: trsfile/standardparameters.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+from enum import Enum
```
Comment (Siebje) [inline trsfile/standardparameters.py:1]: Please add a description to this file, as it may not be evident to a user what 'standard' means.

File: trsfile/parametermap.py [hunk 2]
```diff
@@ -55,13 +164,56 @@ def lock_content(self):
...
+        StandardTraceSetParameters.DISPLAY_HINT_NUM_TRACES_SHOWN: 1,
+        StandardTraceSetParameters.DISPLAY_HINT_TRACES_OVERLAP: False,
+        StandardTraceSetParameters.DISPLAY_HINT_USE_LOG_SCALE: False,
+        StandardTraceSetParameters.X_OFFSET: 0,
+        StandardTraceSetParameters.TRACE_OFFSET: 0,
+        StandardTraceSetParameters.X_SCALE: 1.0,
+        StandardTraceSetParameters.Y_SCALE: 1.0
+    }
+
     def __setitem__(self, key, value):
```
Comment (Siebje) [inline trsfile/parametermap.py:179]: Since we're raising an error if the value is incorrectly typed, a type hint for the argument would be nice.

File: trsfile/parametermap.py [hunk 3]
```diff
@@ -55,13 +164,56 @@ def lock_content(self):
...
+        except TypeError:
+            typed_param = ParameterMapUtil.get_typed_parameter(value)
+        self[name] = typed_param(ParameterMapUtil.to_list_if_listable(value))
+
+    def add_standard_parameter(self, std_trace_set_param: StandardTraceSetParameters, value):
+        """Add a standard trace set parameter with a given value"""
+        typed_param = std_trace_set_param.parameter_type.param_class
+        self[std_trace_set_param.identifier] = typed_param(ParameterMapUtil.to_list_if_listable(value))
+
+    def fill_from_headers(self, headers):
```
Comment (Siebje) [inline trsfile/parametermap.py:202]: please type hint this as well (I'm guessing it's a dict of Header types?)

File: trsfile/parametermap.py [hunk 4]
```diff
@@ -125,6 +277,18 @@ def serialize(self):
             out.extend(value.serialize())
         return out
 
+    @staticmethod
+    def from_trace_parameter_map(trace_parameters):
```
Comment (Siebje) [inline trsfile/parametermap.py:281]: type hints?

File: trsfile/engine/trs.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def __init__(self, path, mode = 'x', **options):
 		headers = options.get('headers', None)
 		self.live_update = int(options.get('live_update', False))
 		self.live_update_count = 0
-		self.padding_mode = options.get('padding_mode', TracePadding.NONE)
+		self.padding_mode = options.get('padding_mode', TracePadding.AUTO)
```
Comment (Siebje) [inline trsfile/engine/trs.py:61]: I'm not sure if this is the correct approach. If I understand it correctly, padding mode auto means that you would need to correct the length of already added traces if you encounter a longer trace added to the set. Since this is not supported by the TRS format, I would shy away from using 'padding mode' at all, or even deprecate it.

What you seem to be doing now is changing what the padding mode argument means, which I doubt is something you want to do.

Comment (Siebje) [reply trsfile/engine/trs.py:61]: After reading more of the code, it seems the use of the padding mode is slightly different than what I thought. The way it's used it maybe actually nice, so if this works, let's document it properly and leave it in.

the way I understand it to work now: If you pre-initialize the length of the trace and data (the full size of the contents of the definitions map), it will automatically pad any shorter entries. Keep in mind that padding a definitions map is not really possible though, and may have unexpected side effects.

File: trsfile/engine/trs.py [hunk 6]
```diff
@@ -380,11 +408,11 @@ def __create_headers(self, headers):
 		# Make sure correct trs version is set
 		if (Header.TRACE_PARAMETER_DEFINITIONS in self.headers or
 			Header.TRACE_SET_PARAMETERS in self.headers) and \
-			(Header.TRS_VERSION not in self.headers or self.headers[Header.TRS_VERSION] < 2):
+				(Header.TRS_VERSION not in self.headers or self.headers[Header.TRS_VERSION] < 2):
 			self.headers[Header.TRS_VERSION] = 2
 
 		# Finally add some extra headers that are freaking useful if they are not provided
```
Comment (Siebje) [inline trsfile/engine/trs.py:414]: maybe reword this to not say 'freaking' lol

File: trsfile/engine/trs.py [hunk 7]
```diff
@@ -393,7 +421,7 @@ def __create_headers(self, headers):
 		# Write the initial headers
 		self.__write_headers()
 
-	def __write_headers(self, headers = None):
+	def __write_headers(self, headers=None):
 		# Write the headers to the file. WARNING: Do not call if the new headers have a different size than the
```
Comment (Siebje) [inline trsfile/engine/trs.py:425]: I would prefer the execution of the code to give this warning to the user rather than a comment in the code.

**Review by Siebje — CHANGES_REQUESTED** [review]
General comments:
- Props for writing comprehensive tests
- Please consider going over your code and adding type hints where possible (don't forget return types)
- Please reformat the files you touch. I noticed that some files are tab indented and some are space indented. Our standard uses spaces, so please correct the files you change.

File: trsfile/engine/trs.py (hunk 5 above)
Comment (TomHogervorst) [reply trsfile/engine/trs.py:61]: You're right, I've removed the code that tries to pad an existing definitions map with data from a trace. It will only add a definition map with none exists yet.

File: trsfile/engine/trs.py (hunk 7 above)
Comment (TomHogervorst) [reply trsfile/engine/trs.py:425]: I've move the check for this from the update_headers method to here.

**Review by Siebje — APPROVED** [review]
